// connect-react-native/lib/module/internal/types.js
export let DapiTheme = /*#__PURE__*/function (DapiTheme) {
  DapiTheme["light"] = "light";
  DapiTheme["dark"] = "dark";
  DapiTheme["dynamic"] = "dynamic";
  return DapiTheme;
}({});
export let DapiLanguage = /*#__PURE__*/function (DapiLanguage) {
  DapiLanguage["en"] = "en";
  DapiLanguage["ar"] = "ar";
  return DapiLanguage;
}({});
export let DapiEnvironment = /*#__PURE__*/function (DapiEnvironment) {
  DapiEnvironment["sandbox"] = "sandbox";
  DapiEnvironment["production"] = "production";
  return DapiEnvironment;
}({});
export let DapiTransactionsType = /*#__PURE__*/function (DapiTransactionsType) {
  DapiTransactionsType["enriched"] = "enriched";
  DapiTransactionsType["categorized"] = "categorized";
  DapiTransactionsType["default"] = "default";
  return DapiTransactionsType;
}({});

/**
 * @deprecated Not usable anymore because Dapi backend accepts the default endpoints paths only now
 */
export let DapiEndpoint = /*#__PURE__*/function (DapiEndpoint) {
  DapiEndpoint["exchangeToken"] = "auth/ExchangeToken";
  DapiEndpoint["getIdentity"] = "data/identity/get";
  DapiEndpoint["getAccounts"] = "data/accounts/get";
  DapiEndpoint["getAccountMetadata"] = "metadata/accounts/get";
  DapiEndpoint["getTransactions"] = "data/transactions/get";
  DapiEndpoint["getBeneficiaries"] = "payment/beneficiaries/get";
  DapiEndpoint["createBeneficiary"] = "payment/beneficiaries/create";
  DapiEndpoint["createTransfer"] = "payment/transfer/create";
  DapiEndpoint["createTransferToExistingBeneficiary"] = "payment/transfer/autoflow";
  DapiEndpoint["getWireBeneficiaries"] = "wire/beneficiaries/get";
  DapiEndpoint["createWireBeneficiary"] = "wire/beneficiaries/create";
  DapiEndpoint["createWireTransfer"] = "wire/transfer/autoflow";
  DapiEndpoint["createWireTransferToExistingBeneficiary"] = "wire/transfer/create";
  DapiEndpoint["delete"] = "users/DelinkUser";
  return DapiEndpoint;
}({});
var TransactionType = /*#__PURE__*/function (TransactionType) {
  TransactionType["CREDIT"] = "credit";
  TransactionType["DEBIT"] = "debit";
  return TransactionType;
}(TransactionType || {});
export let BeneficiaryType = /*#__PURE__*/function (BeneficiaryType) {
  BeneficiaryType["SAME"] = "same";
  BeneficiaryType["LOCAL"] = "local";
  BeneficiaryType["INTERNATIONAL"] = "intl";
  BeneficiaryType["OWN"] = "own";
  return BeneficiaryType;
}({});
